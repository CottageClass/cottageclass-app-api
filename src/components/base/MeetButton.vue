<template>
  <div class="meet-button"
        :class="meetStatus + ' ' + fillStyle + ' ' + layoutStyle + (layoutStyle==='fat' ? ' w-button' : '')"
        @click="meetButtonClick">
    {{meetButtonText}}
    <img v-if="meetStatus==='sent'"
          src="@/assets/check-green.svg"
          alt=""
          class="checkmark-image">
  </div>
</template>

<script>
import { setTimeout, clearTimeout } from 'timers'
import { initProxySession } from '@/utils/api'
import { mapGetters } from 'vuex'
import { alerts, redirect, messaging } from '@/mixins'

export default {
  name: 'MeetButton',
  props: ['targetUser', 'fillStyle', 'layoutStyle'],
  mixins: [alerts, redirect, messaging],
  data () {
    return {
      meetStatus: 'none'
    }
  },
  methods: {
    meetButtonClick (event) {
      this.$emit('meetButtonClick', { event })
      switch (this.meetStatus) {
        case 'none':
          this.showDescriptionModal()
          break
        case 'sending':
          this.undoMessageSending()
          break
      }
    },
    showDescriptionModal () {
      if (this.shouldShowDescriptionModal) {
        this.$store.commit('showModal', {
          modal: {
            title: `We've sent ${this.targetUser.firstName} your request to meet!`,
            bodyText: this.targetUser.firstName + ' will reply if interested. When she does, all you have to do is:<br/>' +
            '1. Decide on a time to meet<br/>' +
            '2. Decide on a place to meet (we recommend meeting in one of your homes)<br/>' +
            '3. Decide on an activity (This is optional. Sometimes itâ€™s best to just meet and talk!)<br/>',
            buttonNames: ['OK'],
            closeCallback: this.checkAuthenticationAndInitiateMessageSending.bind(this)
          }
        })
        if (this.$route.name === 'Events') {
          this.$store.commit('setHasShowEventsPageMessagingDescription')
        }
      } else {
        this.checkAuthenticationAndInitiateMessageSending()
      }
    },
    checkAuthenticationAndInitiateMessageSending () {
      if (this.redirectToSignupIfNotAuthenticated()) {
        this.$store.commit('addPendingWave', { targetUser: this.targetUser })
      } else {
        this.initiateMessageSending()
      }
    },
    undoMessageSending () {
      clearTimeout(this.sendingTimeout)
      this.meetStatus = 'none'
    },
    initiateMessageSending: async function () {
      const that = this
      that.meetStatus = 'sending'
      that.sendingTimeout = setTimeout(function () {
        that.sendMessage()
      }, 4000)
    },
    sendMessage: async function () {
      try {
        await initProxySession(this.currentUser.id,
          this.targetUser.id,
          this.meetMessage(this.targetUser),
          this.acknowledgeMessage(this.targetUser))
        this.meetStatus = 'sent'
        this.$store.commit('addSentWave', { targetUserId: this.targetUser.id })
      } catch (e) {
        console.error(e)
        this.meetStatus = 'none'
        this.showBriefAllert('There was a problem sending your message.  Please try again later', 'failure')
      }
    }
  },
  computed: {
    shouldShowDescriptionModal () {
      if (this.$route.name === 'ProviderProfile') {
        return true
      }
      if (this.$route.name === 'Events' && !this.hasShowEventsPageMessagingDescription) {
        return true
      }
      return false
    },
    meetButtonText () {
      switch (this.meetStatus) {
        case 'none':
          return (this.fillStyle === 'outline') ? 'Wave' : 'Invite for a playdate'
        case 'sending':
          return 'Undo'
        case 'sent':
          return 'Sent'
        default:
          return 'Wave'
      }
    },
    ...mapGetters([ 'currentUser', 'waveHasBeenSent', 'hasShowEventsPageMessagingDescription' ])
  },
  created () {
    if (this.waveHasBeenSent(this.targetUser.id)) {
      this.meetStatus = 'sent'
    }
  }
}
</script>

<style scoped lang="scss">

.meet-button.solid {
  color: white;
  background-color: #1f88e9;

  &.none:hover {
    background-image: linear-gradient(180deg, rgba(0, 0, 0, .02), rgba(0, 0, 0, .04));
  }
  &.sending {
    color: rgb(212, 80, 18);
    border: 1px rgb(212, 80, 18) solid;
    background-color: transparent;
    &:hover {
      background-image: linear-gradient(180deg, rgba(255, 0, 0, .1), rgba(255, 0, 0, .04));
    }
  }
  &.sent {
    color:  rgb(12, 186, 82);
    border-color:  rgb(12, 186, 82);
    background-color:  rgba(12, 186, 82, .2);
    border-width: 2px;
    cursor: default;
  }

}

.meet-button.outline {
  border: 1px solid #1f88e9;
  background-color: transparent;
  color: #1f88e9;
  -webkit-text-fill-color: #1f88e9;  // DO NOT REMOVE.  REQUIRED FOR SAFARI

  &.none:hover {
    background-image: linear-gradient(180deg, rgba(0, 0, 0, .02), rgba(0, 0, 0, .04));
  }
  &.sending {
    color: rgb(212, 80, 18);
    -webkit-text-fill-color: rgb(212, 80, 18);  // DO NOT REMOVE.  REQUIRED FOR SAFARI
    border-color:  rgb(212, 80, 18);

    &:hover {
      background-image: linear-gradient(180deg, rgba(255, 0, 0, .1), rgba(255, 0, 0, .04));
    }
  }
  &.sent {
    color:  rgb(12, 186, 82);
    -webkit-text-fill-color: rgb(12, 186, 82);  // DO NOT REMOVE.  REQUIRED FOR SAFARI
    border-color:  rgb(12, 186, 82);
    background-color:  rgba(12, 186, 82, .2);
    border-width: 2px;
    cursor: default;
  }
}

.meet-button {
  text-align: center;
  border-radius: 4px;
  font-size: 13px;
  &.sent {
    cursor: default;
  }
}
.meet-button.slim {
  display: flex;
  width: 100%;
  margin-bottom: 0px;
  padding: 6px 10px;
  justify-content: center;
  align-items: center;
  font-size: 12px;
  line-height: 15px;
  text-align: center;
}
.meet-button.fat {
    min-width: 80px;
}
</style>
