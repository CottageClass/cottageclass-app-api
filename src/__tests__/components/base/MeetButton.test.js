import { mount } from '@vue/test-utils'
import MeetButton from '@/components/base/MeetButton.vue'

describe('MeetButton', () => {
  let wrapper

  describe('locality provided', () => {
    const targetUser = {
      id: '1230',
      firstName: 'Chuck',
      lastInitial: 'D',
      childAgesInMonths: [1, 15, 99],
      location: {
        lat: 41.969880,
        lng: -73.651964 },
      employer: 'Lockheed',
      jobPosition: 'engineer'
    }
    beforeAll(async () => {
      const $route = {
        name: 'Search'
      }
      const $router = { push: jest.fn() }
      const $store = {
        commit: jest.fn(),
        getters: {
          isAuthenticated: true,
          waveHasBeenSent: () => false,
          currentUser: {
            id: '987654321',
            firstName: 'Natasha',
            childAgesInMonths: [9, 100, 54],
            locality: 'Nashville',
            location: {
              lat: 41.9,
              lng: -73.6 }
          },
          distanceFromCurrentUser () {
            return 4.2
          }
        }
      }
      wrapper = mount(MeetButton, {
        mocks: { $router, $store, $route },
        stubs: ['router-link'],
        propsData: {
          targetUser,
          defaultText: 'Invite',
          showDescriptionModal: false,
          allowUndo: true
        }
      })
    })

    it('displays correct information', () => {
      const meetButton = wrapper.find('.event-action-button')
      expect(meetButton.text()).toEqual('Invite')
      expect(wrapper.vm.meetMessage(targetUser)).toMatchSnapshot()
    })

    it('initiates sending sequence', done => {
      const meetButton = wrapper.find('.event-action-button')
      const mocksendMessage = jest.fn(() => {})
      wrapper.setMethods({ sendMessage: mocksendMessage })
      meetButton.trigger('click')

      // wait a tick so button can update
      wrapper.vm.$nextTick(() => {
        expect(mocksendMessage.mock.calls).toHaveLength(1)
        done()
      })
    })
  })

  describe('no locality provided', () => {
    it('has the correct data', async () => {
      const $router = { push: jest.fn() }
      const $store = {
        commit: jest.fn(),
        getters: {
          waveHasBeenSent: () => false,
          isAuthenticated: true,
          currentUser: {
            id: '987654321',
            firstName: 'Natasha',
            childAgesInMonths: [111, 2, 2],
            locality: '',
            location: {
              lat: 41.9,
              lng: -73.6 }
          },
          distanceFromCurrentUser () {
            return 4.2
          }
        }
      }
      const $route = {
        name: 'Search'
      }

      const targetUser = {
        id: '1230',
        firstName: 'Chuck',
        lastInitial: 'D',
        childAgesInMonths: [1, 2, 3],
        location: {
          lat: 41.969880,
          lng: -73.651964 },
        employer: 'Lockheed',
        jobPosition: 'engineer'
      }
      wrapper = mount(MeetButton, {
        mocks: { $router, $store, $route },
        stubs: ['router-link'],
        propsData: {
          targetUser,
          showDescriptionModal: false,
          defaultText: 'Say Hi',
          allowUndo: false
        }
      })
      const meetButton = wrapper.find('.event-action-button')

      expect(meetButton.text()).toEqual('Say Hi')
      expect(wrapper.vm.meetMessage(targetUser)).toMatchSnapshot()
    })
  })

  describe('negative age kid', () => {
    const targetUser = {
      id: '1230',
      firstName: 'Chuck',
      lastInitial: 'D',
      childAgesInMonths: [1, 2, 3],
      location: {
        lat: 41.969880,
        lng: -73.651964 },
      employer: 'Lockheed',
      jobPosition: 'engineer'
    }

    beforeAll(async () => {
      const $router = { push: jest.fn() }
      const $store = {
        commit: jest.fn(),
        getters: {
          waveHasBeenSent: () => false,
          isAuthenticated: true,
          currentUser: {
            id: '987654321',
            firstName: 'Natasha',
            childAgesInMonths: [-1],
            locality: 'Nashville',
            location: {
              lat: 41.9,
              lng: -73.6 }
          },
          distanceFromCurrentUser () {
            return 4.2
          }
        }
      }
      const $route = {
        name: 'Search'
      }
      wrapper = mount(MeetButton, {
        mocks: { $router, $store, $route },
        stubs: ['router-link'],
        propsData: {
          targetUser,
          showDescriptionModal: false,
          defaultText: 'Wave',
          allowUndo: true
        }
      })
    })

    it('displays correct information', () => {
      const meetButton = wrapper.find('.event-action-button')
      expect(meetButton.text()).toEqual('Wave')
      expect(wrapper.vm.meetMessage(targetUser)).toMatchSnapshot()
    })

    it('initiates sending sequence', done => {
      const meetButton = wrapper.find('.event-action-button')
      const mocksendMessage = jest.fn(() => {})
      wrapper.setMethods({ sendMessage: mocksendMessage })
      meetButton.trigger('click')

      // wait a tick so button can update
      wrapper.vm.$nextTick(() => {
        expect(mocksendMessage.mock.calls).toHaveLength(1)
        done()
      })
    })
  })
})
