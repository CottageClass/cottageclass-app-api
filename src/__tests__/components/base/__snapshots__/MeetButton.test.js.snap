// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetButton locality provided displays correct information 1`] = `"Natasha (http://localhost/users/987654321 ) waved at you! They live 4.2 mi. away with 3 kids age 9 mos, 4 & 8. If you're interested in a playdate, say hi!"`;

exports[`MeetButton negative age kid displays correct information 1`] = `"Natasha (http://localhost/users/987654321 ) waved at you! They live 4.2 mi. away. If you're interested in a playdate, say hi!"`;

exports[`MeetButton no locality provided has the correct data 1`] = `"Natasha (http://localhost/users/987654321 ) waved at you! They live 4.2 mi. away with 3 kids age 2 mos, 2 mos & 9. If you're interested in a playdate, say hi!"`;
